src/ring_theory/integral_closure.lean

```diff
@@ -242,6 +243,38 @@ lemma integral_closure.is_integral (x : integral_closure R A) : is_integral R x
 let ⟨p, hpm, hpx⟩ := x.2 in ⟨p, hpm, subtype.eq $
 by rwa [subtype.val_eq_coe, ← subalgebra.val_apply, aeval_alg_hom_apply] at hpx⟩
 
+theorem is_integral_of_is_integral_mul_unit {x y : A} {r : R} (hr : (algebra_map R A r) * y = 1)
+  (hx : is_integral R (x * y)) : (is_integral R x) :=
+begin
+  by_cases triv : (1 : R) = 0,
+  { exact ⟨0, ⟨trans leading_coeff_zero triv.symm, eval₂_zero _ _⟩⟩ },
+  haveI : nontrivial R := nontrivial_of_ne 1 0 triv,
+  obtain ⟨p, ⟨p_monic, hp⟩⟩ := hx,
+  let p' : polynomial R := ∑ (i : ℕ) in p.support, monomial i ((p.coeff i) * r ^ (p.nat_degree - i)),
+  have hp' : p'.nat_degree = p.nat_degree,
+  { refine finset_sum_monomial_degree (λ h, triv _),
+    rw [nat.sub_self, pow_zero r, mul_one] at h,
+    exact p_monic ▸ h },
+  refine ⟨p', ⟨_, _⟩⟩,
+  { rwa [monic, leading_coeff, hp', finset_sum_monomial_coeff, nat.sub_self, pow_zero, mul_one] },
+  -- TODO: A lot of this is just basic algebra, but `ring` can't solve it because only some elements are invertible
+  { have : is_unit (y ^ p.nat_degree) := is_unit_iff_exists_inv.2
+      ⟨(algebra_map R A r) ^ p.nat_degree, by rw [← mul_pow, mul_comm, hr, one_pow]⟩,
+    rw [aeval_def, eval₂, sum_over_range] at hp,
+    rw [← (is_unit.mul_right_eq_zero this), aeval_def, eval₂, sum_over_range, finset.mul_sum, ← hp],
+    refine finset.sum_congr (by rw hp') (λ n hn, _),
+    simp only [finset_sum_monomial_coeff, ring_hom.map_pow, ring_hom.map_mul],
+    rw [← mul_assoc (y ^ p.nat_degree), mul_comm (y ^ p.nat_degree), mul_comm x y, mul_pow y x n,
+        ← mul_assoc _ (y ^ n) (x ^ n), mul_assoc _ _ (y ^ p.nat_degree)],
+    congr,
+    have : is_unit ((algebra_map R A r) ^ n) := is_unit_iff_exists_inv.2
+      ⟨y ^ n, by rw [← mul_pow, hr, one_pow]⟩,
+    rw [← (is_unit.mul_left_inj this), ← mul_pow y _ n, mul_comm y, hr, mul_comm, ← mul_assoc,
+      ← pow_add, nat.add_sub_cancel' (nat.le_of_lt_succ (finset.mem_range.1 hn)),
+      ← mul_pow, hr, one_pow, one_pow],
+    all_goals {simp [add_mul]} },
+end
```

Comment:
I think you can shorten the proof dramatically if you use `scale_roots p (r⁻¹)` defined in `ring_theory.polynomial.rational_root`.

Reply:
This is very helpful as that is essentially the exact same polynomial as I tried to use here. The only issue is that if I try and import it, I end up with a cyclic dependency, `rational_root  → localization  → integral_closure  → rational_root`, and I can't just move this theorem to `rational_root` since the additions to `localization` depend on it. I could maybe move these new theorems into `rational_root`, or create a new file like `localization_integral`. Not sure what's the best practice here

Reply:
I think the best way would be to move the whole `section scale_roots` from the `rational_root` file to a new file `ring_theory/polynomial/scale_roots.lean` (or add it to `ring_theory/basic`, but that's getting long by itself already) (or even move it to `data/polynomial/scale_roots.lean`. I'm not sure what qualifies as `data/polynomial` and what as `ring_theory/polynomial`).

Reply:
I've pulled out `scale_roots` into its own file, but to completely resolve the dependencies I also had to move `non_zero_divisors` into its own file, since `scale_roots` depends on it (I managed to remove `le_non_zero_divisors` from some `scale_roots` lemmas that didn't actually need to rely on it, but some do still actually need it`).

 For now these are both in two new independent files, but perhaps `non_zero_divisors` could go somewhere else, maybe `integral_domain` or some other place that already deals with zero divisors. I also kept scale roots in `ring_theory` rather than `data` because of the reliance on `non_zero_divisors` for some lemmas.

Reply:
Aha, I recall being annoyed before that `non_zero_divisors` is so far down the import hierarchy. Thank you for fixing that :tada;

As for the location of `non_zero_divisors`, maybe we should generalize all the way to `monoid_with_zero` and put it somewhere in `algebra/group_with_zero.lean` (I can't think of any non-`ring` `monoid_with_zero` where we would care about zero divisors, but the highest possible generality is what we really care about in mathlib :P) 

I don't have any time right now for a full review, but tomorrow morning (in about 9 hours) I will check it out :+1:

Reply:
I changed the definition to the more generic monoid with zero, but most of the useful lemmas require more structure to prove, so as of now they still rely on either `comm_ring` or `integral_domain`. I didn't move it to location outside `ring_theory` either, since basically everything but the definition still relies on `comm_ring`